source/common/config/version_converter.cc

```diff
@@ -1,43 +1,140 @@
 #include "common/config/version_converter.h"
 
+#include "common/common/assert.h"
 #include "common/config/api_type_oracle.h"
+#include "common/protobuf/well_known.h"
 
 namespace Envoy {
 namespace Config {
 
 namespace {
 
+void messageFieldVisitor(std::function<void(Protobuf::Message&, const Protobuf::FieldDescriptor&,
+                                            const Protobuf::Reflection&)>
+                             field_fn,
+                         Protobuf::Message& message) {
+  const Protobuf::Descriptor* descriptor = message.GetDescriptor();
+  const Protobuf::Reflection* reflection = message.GetReflection();
+  for (int i = 0; i < descriptor->field_count(); ++i) {
+    const Protobuf::FieldDescriptor* field = descriptor->field(i);
+    field_fn(message, *field, *reflection);
+    // If this is a message, recurse to scrub deprecated fields in the sub-message.
+    if (field->cpp_type() == Protobuf::FieldDescriptor::CPPTYPE_MESSAGE) {
+      if (field->is_repeated()) {
+        const int size = reflection->FieldSize(message, field);
+        for (int j = 0; j < size; ++j) {
+          messageFieldVisitor(field_fn, *reflection->MutableRepeatedMessage(&message, field, j));
+        }
+      } else if (reflection->HasField(message, field)) {
+        messageFieldVisitor(field_fn, *reflection->MutableMessage(&message, field));
+      }
+    }
+  }
+}
+
 // Reinterpret a Protobuf message as another Protobuf message by converting to
 // wire format and back. This only works for messages that can be effectively
 // duck typed this way, e.g. with a subtype relationship modulo field name.
 void wireCast(const Protobuf::Message& src, Protobuf::Message& dst) {
   dst.ParseFromString(src.SerializeAsString());
 }
 
+// Create a new dynamic message based on some message wire cast to the target
+// descriptor. If the descriptor is null, a copy is performed.
+DynamicMessagePtr createForDescriptorWithCast(const Protobuf::Message& message,
+                                              const Protobuf::Descriptor* desc) {
+  auto dynamic_message = std::make_unique<DynamicMessage>();
+  if (desc != nullptr) {
+    dynamic_message->msg_.reset(dynamic_message->dynamic_msg_factory_.GetPrototype(desc)->New());
+    wireCast(message, *dynamic_message->msg_);
```

Comment:
Should wireCast check for error and/or RELEASE_ASSERT? It seems bad that we don't do that right now, orthogonal to this PR.

Reply:
Yeah, that's fine to do; we should only ever encounter this if there is an internal programming error, so let's make it an assertion.